File: 18. 4Sum/18. 4Sum.go
func fourSum(nums []int, target int) [][]int {
    sort.Ints(nums)
    result := [][]int{}
    
    for a := 0; a < len(nums) - 3; a++ {
        if a > 0 && nums[a] == nums[a-1] {
            continue
        }
        for b := a + 1; b < len(nums) - 2; b++ {
            if b > a + 1 && nums[b] == nums[b-1] {
                continue
            }
            c, d := b + 1, len(nums) - 1
            for c < d {
                sum := nums[a] + nums[b] + nums[c] + nums[d]
                if sum == target {
                    result = append(result, []int{nums[a], nums[b], nums[c], nums[d]})
                    for c < d && nums[c] == nums[c+1] {
                        c++
                    }
                    for c < d && nums[d] == nums[d-1] {
                        d--
                    }
                    c++
                    d--
                } else if sum < target {
                    c++
                } else {
                    d--
                }
            }
        }
    }
    
    return result
}

File: 18. 4Sum/18. 4Sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
    nums.sort!
    result = []
    
    (0..nums.length - 4).each do |a|
        next if a > 0 && nums[a] == nums[a-1]
        (a+1..nums.length - 3).each do |b|
            next if b > a + 1 && nums[b] == nums[b-1]
            c = b + 1
            d = nums.length - 1
            while c < d
                sum = nums[a] + nums[b] + nums[c] + nums[d]
                if sum == target
                    result << [nums[a], nums[b], nums[c], nums[d]]
                    c += 1 while c < d && nums[c] == nums[c+1]
                    d -= 1 while c < d && nums[d] == nums[d-1]
                    c += 1
                    d -= 1
                elsif sum < target
                    c += 1
                else
                    d -= 1
                end
            end
        end
    end
    
    result
end